MazerPlatformer/GameWorld.cs

```diff
@@ -145,61 +148,88 @@ public void Initialize()
                 gameObject.Value.OnGameObjectComponentChanged += ValueOfGameObjectComponentChanged; // be informed about this objects component update
                 
                 // Every object will have access to the player
-                gameObject.Value.Components.Add(new Component(Component.ComponentType.Player, _level.Player));
+                gameObject.Value.Components.Add(new Component(Component.ComponentType.Player, Level.Player));
+                
                 // every object will have access to the game world
                 gameObject.Value.Components.Add(new Component(Component.ComponentType.GameWorld, this));
             }
+
+            return Nothing; // TODO: FIXME
         }
         
-        private void PlayerOnOnCollision(GameObject thePlayer, GameObject otherObject)
+        private Either<IFailure, Unit> PlayerOnOnCollision(GameObject thePlayer, GameObject otherObject)
         {
-
             // Change my health component to be affected by the hit points of the other object
-            if (otherObject.Type != GameObjectType.Npc) return;
-
-            var npcTypeComponent = otherObject.FindComponentByType(Component.ComponentType.NpcType);
-            var npcType = (Npc.NpcTypes)npcTypeComponent.Value;
-
-            if (npcType == Npc.NpcTypes.Enemy)
+            return otherObject.Type != GameObjectType.Npc
+                ? Nothing
+                : otherObject.FindComponentByType(Component.ComponentType.NpcType)
+                    .ToEither(NotFound.Create($"Could not find component of type {Component.ComponentType.NpcType} on other object"))
+                    .EnsuringMap(component => (Npc.NpcTypes) component.Value)
```

Comment:
Work of art. This will ensure a bind does not throw exceptions and if any exceptions are thrown the failure is reported as a TransformationFailure. Outstanding!